presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -183,6 +184,155 @@ else if (rescaleFactor > 0) {
         }
     }
 
+    public static Slice add(Slice left, Slice right)
+    {
+        Slice result = unscaledDecimal();
+
+        if (!(isNegative(left) ^ isNegative(right))) {
+            // either both negative or both positive
+            addUnsigned(left, right, result, isNegative(left));
+        }
+        else {
+            int compare = compareUnsigned(left, right);
+            if (compare == 0) {
+                return result;
+            }
+            if (isNegative(left)) {
+                if (compare > 0) {
+                    subtractUnsigned(left, right, result, true);
+                }
+                else {
+                    subtractUnsigned(right, left, result, false);
+                }
+            }
+            else {
+                if (compare > 0) {
+                    subtractUnsigned(left, right, result, false);
+                }
+                else {
+                    subtractUnsigned(right, left, result, true);
+                }
+            }
+        }
+        return result;
+    }
+
+    public static void addUnsigned(Slice left, Slice right, Slice result, boolean resultNegative)
+    {
+        int l0 = getInt(left, 0);
+        int l1 = getInt(left, 1);
+        int l2 = getInt(left, 2);
+        int l3 = getInt(left, 3);
+
+        int r0 = getInt(right, 0);
+        int r1 = getInt(right, 1);
+        int r2 = getInt(right, 2);
+        int r3 = getInt(right, 3);
+
+        long product;
+        product = (l0 & LONG_MASK) + (r0 & LONG_MASK);
+
+        int z0 = (int) product;
+
+        product = (l1 & LONG_MASK) + (r1 & LONG_MASK) + (product >> 32);
+
+        int z1 = (int) product;
+
+        product = (l2 & LONG_MASK) + (r2 & LONG_MASK) + (product >> 32);
+
+        int z2 = (int) product;
+
+        product = (l3 & LONG_MASK) + (r3 & LONG_MASK) + (product >> 32);
+
+        int z3 = (int) product;
+
+        pack(result, z0, z1, z2, z3, resultNegative);
+
+        throwIfOverflows(z0, z1, z2, z3);
```

Comment:
I thought we decided that `throwIfOverflows` should be called explicitly by operators.